### PR TITLE
Port build workspace script to Rust

### PR DESCRIPTION
**Pull request by harryscholes**
Seems a bit unnecessary to have Python as a dependency of for the build workspace script, so I ported it to Rust. The image size is reduced by ~45 MB:

```
❯ docker image ls | head -3
REPOSITORY                         TAG                IMAGE ID       CREATED          SIZE
cosmwasm/workspace-optimizer       0.12.2             fd685a78ecf2   48 seconds ago   925MB
cosmwasm/workspace-optimizer       0.12.2_rust_port   e914e7830468   2 minutes ago    879MB
```

<details>
  <summary>Comparison of 0.12.2_rust_port and 0.12.2 output, showing that the same binaries are produced</summary>

```
❯ docker run --rm -v "$(pwd)":/code \
  --mount type=volume,source="$(basename "$(pwd)")_cache",target=/code/target \
  --mount type=volume,source=registry_cache,target=/usr/local/cargo/registry \
  cosmwasm/workspace-optimizer:0.12.2_rust_port
1.54.0-aarch64-unknown-linux-musl (default)
cargo 1.54.0 (5ae8d74b3 2021-06-22)
Building artifacts in workspace ...
Found workspace member entries: ["packages/*", "contracts/*"]
Package directories: ["contracts/terraswap_factory", "contracts/terraswap_pair", "contracts/terraswap_router", "contracts/terraswap_token", "packages/terraswap"]
Contracts to be built: ["contracts/terraswap_factory", "contracts/terraswap_pair", "contracts/terraswap_router", "contracts/terraswap_token"]
Building "contracts/terraswap_factory" ...
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /code/contracts/terraswap_factory/Cargo.toml
workspace: /code/Cargo.toml
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /code/contracts/terraswap_pair/Cargo.toml
workspace: /code/Cargo.toml
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /code/contracts/terraswap_router/Cargo.toml
workspace: /code/Cargo.toml
    Finished release [optimized] target(s) in 0.91s
Building "contracts/terraswap_pair" ...
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /code/contracts/terraswap_factory/Cargo.toml
workspace: /code/Cargo.toml
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /code/contracts/terraswap_pair/Cargo.toml
workspace: /code/Cargo.toml
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /code/contracts/terraswap_router/Cargo.toml
workspace: /code/Cargo.toml
   Compiling terraswap-pair v1.2.0 (/code/contracts/terraswap_pair)
    Finished release [optimized] target(s) in 4.19s
Building "contracts/terraswap_router" ...
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /code/contracts/terraswap_factory/Cargo.toml
workspace: /code/Cargo.toml
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /code/contracts/terraswap_pair/Cargo.toml
workspace: /code/Cargo.toml
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /code/contracts/terraswap_router/Cargo.toml
workspace: /code/Cargo.toml
   Compiling terraswap-router v1.2.0 (/code/contracts/terraswap_router)
    Finished release [optimized] target(s) in 4.57s
Building "contracts/terraswap_token" ...
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /code/contracts/terraswap_factory/Cargo.toml
workspace: /code/Cargo.toml
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /code/contracts/terraswap_pair/Cargo.toml
workspace: /code/Cargo.toml
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /code/contracts/terraswap_router/Cargo.toml
workspace: /code/Cargo.toml
   Compiling terraswap-token v1.2.0 (/code/contracts/terraswap_token)
    Finished release [optimized] target(s) in 4.21s
Creating intermediate hashes ...
d802f3d64578aefd26242cd2b8e06bb77f064111375056c0ab166ac49e60a1d3  target/wasm32-unknown-unknown/release/terraswap_factory.wasm
6974f36e514ca19cbe07b8b0fe290c982027b613ead4f2b8f9aca69208477624  target/wasm32-unknown-unknown/release/terraswap_pair.wasm
47c347eacd1902dc24a59dae5e74901d6fdfbaae6a981cafe2325e454594baab  target/wasm32-unknown-unknown/release/terraswap_router.wasm
9c63a3edc2be8e6dc91ea29e7db0265b13f768575da2caa65ea5b45701290f0f  target/wasm32-unknown-unknown/release/terraswap_token.wasm
Optimizing artifacts in workspace ...
Optimizing ../target/wasm32-unknown-unknown/release/terraswap_factory.wasm ...
Optimizing ../target/wasm32-unknown-unknown/release/terraswap_pair.wasm ...
Optimizing ../target/wasm32-unknown-unknown/release/terraswap_router.wasm ...
Optimizing ../target/wasm32-unknown-unknown/release/terraswap_token.wasm ...
Moving wasm files ...
Post-processing artifacts in workspace ...
a25ad28b91754a99cfeadcb8f7f6f0570a118378cbd284938620089b4ada5268  terraswap_factory.wasm
8c1ffa6dae8c769454159fb4ca7aeb218d64473b1ce7cf733a2bcfd917d2c499  terraswap_pair.wasm
d9ab1af0091cc4881ed46f56e5858657d92a17e4ec95af79f97a515871ad8fdd  terraswap_router.wasm
9004e977a94804adbc1863217805e4e5c302eca0dcfa6dcfdd0e437a9a7108d2  terraswap_token.wasm
done
❯ docker run --rm -v "$(pwd)":/code \
  --mount type=volume,source="$(basename "$(pwd)")_cache",target=/code/target \
  --mount type=volume,source=registry_cache,target=/usr/local/cargo/registry \
  cosmwasm/workspace-optimizer:0.12.2
1.54.0-aarch64-unknown-linux-musl (default)
cargo 1.54.0 (5ae8d74b3 2021-06-22)
Building artifacts in workspace ...
Found workspace member entries: ['packages/*', 'contracts/*']
Package directories: ['contracts/terraswap_factory', 'contracts/terraswap_pair', 'contracts/terraswap_router', 'contracts/terraswap_token', 'packages/terraswap']
Contracts to be built: ['contracts/terraswap_factory', 'contracts/terraswap_pair', 'contracts/terraswap_router', 'contracts/terraswap_token']
Building contracts/terraswap_factory ...
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /code/contracts/terraswap_factory/Cargo.toml
workspace: /code/Cargo.toml
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /code/contracts/terraswap_pair/Cargo.toml
workspace: /code/Cargo.toml
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /code/contracts/terraswap_router/Cargo.toml
workspace: /code/Cargo.toml
    Finished release [optimized] target(s) in 0.15s
Building contracts/terraswap_pair ...
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /code/contracts/terraswap_factory/Cargo.toml
workspace: /code/Cargo.toml
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /code/contracts/terraswap_pair/Cargo.toml
workspace: /code/Cargo.toml
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /code/contracts/terraswap_router/Cargo.toml
workspace: /code/Cargo.toml
    Finished release [optimized] target(s) in 0.11s
Building contracts/terraswap_router ...
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /code/contracts/terraswap_factory/Cargo.toml
workspace: /code/Cargo.toml
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /code/contracts/terraswap_pair/Cargo.toml
workspace: /code/Cargo.toml
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /code/contracts/terraswap_router/Cargo.toml
workspace: /code/Cargo.toml
    Finished release [optimized] target(s) in 0.12s
Building contracts/terraswap_token ...
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /code/contracts/terraswap_factory/Cargo.toml
workspace: /code/Cargo.toml
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /code/contracts/terraswap_pair/Cargo.toml
workspace: /code/Cargo.toml
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /code/contracts/terraswap_router/Cargo.toml
workspace: /code/Cargo.toml
    Finished release [optimized] target(s) in 0.10s
Creating intermediate hashes ...
d802f3d64578aefd26242cd2b8e06bb77f064111375056c0ab166ac49e60a1d3  target/wasm32-unknown-unknown/release/terraswap_factory.wasm
6974f36e514ca19cbe07b8b0fe290c982027b613ead4f2b8f9aca69208477624  target/wasm32-unknown-unknown/release/terraswap_pair.wasm
47c347eacd1902dc24a59dae5e74901d6fdfbaae6a981cafe2325e454594baab  target/wasm32-unknown-unknown/release/terraswap_router.wasm
9c63a3edc2be8e6dc91ea29e7db0265b13f768575da2caa65ea5b45701290f0f  target/wasm32-unknown-unknown/release/terraswap_token.wasm
Optimizing artifacts in workspace ...
Optimizing ../target/wasm32-unknown-unknown/release/terraswap_factory.wasm ...
Optimizing ../target/wasm32-unknown-unknown/release/terraswap_pair.wasm ...
Optimizing ../target/wasm32-unknown-unknown/release/terraswap_router.wasm ...
Optimizing ../target/wasm32-unknown-unknown/release/terraswap_token.wasm ...
Moving wasm files ...
Post-processing artifacts in workspace ...
a25ad28b91754a99cfeadcb8f7f6f0570a118378cbd284938620089b4ada5268  terraswap_factory.wasm
8c1ffa6dae8c769454159fb4ca7aeb218d64473b1ce7cf733a2bcfd917d2c499  terraswap_pair.wasm
d9ab1af0091cc4881ed46f56e5858657d92a17e4ec95af79f97a515871ad8fdd  terraswap_router.wasm
9004e977a94804adbc1863217805e4e5c302eca0dcfa6dcfdd0e437a9a7108d2  terraswap_token.wasm
done
```
</details>

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 DOCKER_NAME_RUST_OPTIMIZER := "cosmwasm/rust-optimizer"
 DOCKER_NAME_WORKSPACE_OPTIMIZER := "cosmwasm/workspace-optimizer"
-DOCKER_TAG := 0.12.2
+DOCKER_TAG := 0.12.2_rust_port
```
Comment (maurolacy) [inline Makefile:5]: Let's just use `0.12.3` here for consistency. `0.12.2` is still unreleased, but still.

File: build_workspace/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    let file = fs::read_to_string("Cargo.toml").unwrap();
+    let cargo_toml: CargoToml = toml::from_str(&file).unwrap();
+    let members = cargo_toml.workspace.members;
+
+    println!("Found workspace member entries: {:?}", members);
+
+    let mut all_packages = members
+        .iter()
+        .map(|member| glob(member).unwrap().map(|path| path.unwrap()))
+        .flatten()
```
Comment (maurolacy) [inline build_workspace/src/main.rs:31]: Nice.

**Review by maurolacy — APPROVED** [review]
I like it! I will happily merge this, but let's wait for @webmaster128 opinion / review.

Please add a CHANGELOG entry with this change.

File: Makefile (hunk 1 above)
Comment (harryscholes) [reply Makefile:5]: Sure. This was just a temporary version number for demonstration :)

File: build_workspace/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,11 @@
+[package]
+name = "build_workspace"
+version = "0.1.0"
+edition = "2018"
```
Comment (webmaster128) [inline build_workspace/Cargo.toml:4]:
```suggestion
edition = "2018"
publish = false
```

File: Dockerfile [hunk 4]
```diff
@@ -65,8 +65,12 @@ RUN chmod +x /usr/local/bin/optimize.sh
 
 ADD optimize_workspace.sh /usr/local/bin/
 RUN chmod +x /usr/local/bin/optimize_workspace.sh
-ADD build_workspace.py /usr/local/bin/
-RUN chmod +x /usr/local/bin/build_workspace.py
+
+ADD build_workspace build_workspace
+ENV RUSTFLAGS="-C link-arg=-s"
```
Comment (webmaster128) [inline Dockerfile:70]: Is this really necessary? How large is `build_workspace`? I'd prefer sticking to solid defaults for native builds.

**Review by webmaster128 — COMMENTED** [review]
Very nice idea, thank you for this work!

File: Dockerfile (hunk 4 above)
Comment (harryscholes) [reply Dockerfile:70]: 500K vs 688K, so no it's not really necessary :) Will remove

Comment (webmaster128) [reply Dockerfile]: Thanks. Not worth thinking about it in that case.

**Pull request merged**